From c62b47aab2a79eb14b806934cd8916134dad069d Mon Sep 17 00:00:00 2001
From: Valentin Raevsky <valentin@compulab.co.il>
Date: Tue, 2 Feb 2021 15:36:21 +0200
Subject: [PATCH] brcmfmac: Add DFS channel enable/disable option

Added a driver read only parameter: brcmf_dfs_disable.
Values:
0 - DFS channels 52-148 enabled
1 - DFS channels 52-148 disabled (default)

Signed-off-by: Valentin Raevsky <valentin@compulab.co.il>
---
 .../broadcom/brcm80211/brcmfmac/cfg80211.c         | 31 +++++++++++++++++-----
 .../wireless/broadcom/brcm80211/brcmfmac/common.c  |  4 +++
 2 files changed, 29 insertions(+), 6 deletions(-)

diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c
index 04a58aa2a571..ae2f941334fe 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c
@@ -30,6 +30,8 @@
 #include "bus.h"
 #include "common.h"
 
+extern int brcmf_dfs_disable;
+
 #define BRCMF_SCAN_IE_LEN_MAX		2048
 
 #define WPA_OUI				"\x00\x50\xF2"	/* WPA OUI */
@@ -155,6 +157,12 @@ static struct ieee80211_channel __wl_5ghz_channels[] = {
 	CHAN5G(153), CHAN5G(157), CHAN5G(161), CHAN5G(165)
 };
 
+static struct ieee80211_channel __wl_5ghz_channels_dfs_disabled[] = {
+	CHAN5G(34), CHAN5G(36), CHAN5G(38), CHAN5G(40), CHAN5G(42),
+	CHAN5G(44), CHAN5G(46), CHAN5G(48), CHAN5G(149),
+	CHAN5G(153), CHAN5G(157), CHAN5G(161), CHAN5G(165)
+};
+
 /* Band templates duplicated per wiphy. The channel info
  * above is added to the band during setup.
  */
@@ -6280,8 +6288,9 @@ static int brcmf_construct_chaninfo(struct brcmf_cfg80211_info *cfg,
 			/* It seems firmware supports some channel we never
 			 * considered. Something new in IEEE standard?
 			 */
-			bphy_err(drvr, "Ignoring unexpected firmware channel %d\n",
-				 ch.control_ch_num);
+			if (!brcmf_dfs_disable)
+				bphy_err(drvr, "Ignoring unexpected firmware channel %d\n",
+					 ch.control_ch_num);
 			continue;
 		}
 
@@ -6925,15 +6934,25 @@ static int brcmf_setup_wiphy(struct wiphy *wiphy, struct brcmf_if *ifp)
 			if (!band)
 				return -ENOMEM;
 
-			band->channels = kmemdup(&__wl_5ghz_channels,
-						 sizeof(__wl_5ghz_channels),
-						 GFP_KERNEL);
+			if (brcmf_dfs_disable)
+				band->channels = kmemdup(&__wl_5ghz_channels_dfs_disabled,
+							 sizeof(__wl_5ghz_channels_dfs_disabled),
+							 GFP_KERNEL);
+			else
+				band->channels = kmemdup(&__wl_5ghz_channels,
+							 sizeof(__wl_5ghz_channels),
+							 GFP_KERNEL);
+
 			if (!band->channels) {
 				kfree(band);
 				return -ENOMEM;
 			}
 
-			band->n_channels = ARRAY_SIZE(__wl_5ghz_channels);
+			if (brcmf_dfs_disable)
+				band->n_channels = ARRAY_SIZE(__wl_5ghz_channels_dfs_disabled);
+			else
+				band->n_channels = ARRAY_SIZE(__wl_5ghz_channels);
+
 			wiphy->bands[NL80211_BAND_5GHZ] = band;
 		}
 	}
diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.c
index f1b9faafdda1..feb1046a243d 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.c
@@ -43,6 +43,10 @@ int brcmf_msg_level;
 module_param_named(debug, brcmf_msg_level, int, 0600);
 MODULE_PARM_DESC(debug, "Level of debug output");
 
+int brcmf_dfs_disable=1;
+module_param_named(dfs_disable, brcmf_dfs_disable, int, 0400);
+MODULE_PARM_DESC(dfs_disable, "Disable DFS channels");
+
 static int brcmf_p2p_enable;
 module_param_named(p2pon, brcmf_p2p_enable, int, 0);
 MODULE_PARM_DESC(p2pon, "Enable legacy p2p management functionality");
-- 
2.11.0

